### PR TITLE
feat(*): add routes v2 functionality

### PR DESCRIPTION
**Pull request by michelleN**
This programs envoy proxies with new route refactor
based inbound/outbound route calculations behind a feature
flag.

Signed-off-by: Michelle Noorali <minooral@microsoft.com>

**Description**:

**Affected area**:

- New Functionality      [ ]
- Documentation          [ ]
- Install                [ ]
- Control Plane          [ ]
- CLI Tool               [ ]
- Certificate Management [ ]
- Networking             [ ]
- Metrics                [ ]
- SMI Policy             [ ]
- Security               [ ]
- Tests                  [ ]
- CI System              [ ]
- Performance            [ ]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?

### CODE REVIEW

File: experimental/routes_refactor/demo/manifests/traffic-access.yaml [hunk 1]
```diff
@@ -57,4 +57,35 @@ spec:
   sources:
```
Comment (snehachhabria) [inline experimental/routes_refactor/demo/manifests/traffic-access.yaml:57]: If I understand this right, the refactor doesn't support traffic split yet ?

File: experimental/routes_refactor/demo/scripts/port-forward-bookbuyer-envoy.sh [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
```
Comment (snehachhabria) [inline experimental/routes_refactor/demo/scripts/port-forward-bookbuyer-envoy.sh:1]: Why are we adding port forwarding scripts again done these already exist in `./scripts`

File: pkg/envoy/rds/new_response.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	"github.com/openservicemesh/osm/pkg/catalog"
+	cat "github.com/openservicemesh/osm/pkg/catalog"
+	"github.com/openservicemesh/osm/pkg/envoy"
+	"github.com/openservicemesh/osm/pkg/envoy/route"
+)
+
+func newResponse(catalog catalog.MeshCataloger, proxy *envoy.Proxy) (*xds_discovery.DiscoveryResponse, error) {
+	proxyIdentity, err := cat.GetServiceAccountFromProxyCertificate(proxy.GetCertificateCommonName())
+	if err != nil {
+		log.Error().Err(err).Msgf("Error looking up Service Account for Envoy with CN=%q", proxy.GetCertificateCommonName())
```
Comment (snehachhabria) [inline pkg/envoy/rds/new_response.go:16]: Based on the latest changes in logs for audit, we no longer want to log the certificate common name, we should be logging the serial number

File: pkg/envoy/rds/new_response.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+	inboundTrafficPolicies, outboundTrafficPolicies, err := catalog.ListTrafficPoliciesForServiceAccount(proxyIdentity)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := &xds_discovery.DiscoveryResponse{
+		TypeUrl: string(envoy.TypeRDS),
+	}
+
+	// TODO merge ingress policies with existing inboundTrafficPolicies
```
Comment (snehachhabria) [inline pkg/envoy/rds/new_response.go:28]: Could we create an issue for this TODO and add the issue number here

File: pkg/trafficpolicy/trafficpolicy.go [hunk 5]
```diff
@@ -33,6 +33,16 @@ func NewOutboundTrafficPolicy(name string, hostnames []string) *OutboundTrafficP
 	}
 }
 
+// TotalClustersWeight returns total weight of the WeightedClusters in trafficpolicy.RouteWeightedClusters
+func (rwc *RouteWeightedClusters) TotalClustersWeight() int {
```
Comment (snehachhabria) [inline pkg/trafficpolicy/trafficpolicy.go:37]: A unit test for this method too would be great

File: experimental/routes_refactor/demo/manifests/traffic-access.yaml (hunk 1 above)
Comment (michelleN) [reply experimental/routes_refactor/demo/manifests/traffic-access.yaml:57]: That's correct. @ksubrmnn will work on the traffic split piece once this is merged.

File: demo/run-osm-demo.sh [hunk 6]
```diff
@@ -36,6 +36,8 @@ DEPLOY_PROMETHEUS="${DEPLOY_PROMETHEUS:-true}"
 ENABLE_PROMETHEUS_SCRAPING="${ENABLE_PROMETHEUS_SCRAPING:-true}"
 DEPLOY_WITH_SAME_SA="${DEPLOY_WITH_SAME_SA:-false}"
 ENVOY_LOG_LEVEL="${ENVOY_LOG_LEVEL:-debug}"
+ROUTES_V2="${ROUTES_V2:-false}"
```
[the remark this answers is not in the document]
Comment (michelleN) [reply demo/run-osm-demo.sh:39]: yes! Thank you. Will add.

File: experimental/routes_refactor/demo/scripts/port-forward-bookbuyer-envoy.sh (hunk 2 above)
Comment (michelleN) [reply experimental/routes_refactor/demo/scripts/port-forward-bookbuyer-envoy.sh:1]: These port-forwarding scripts are slightly different as they use an additional label selector (version) for the bookstore pods. It seemed convenient to put all the relevant scripts in one place (in the experimental directory) to easy access them while running the routes refactor demo in the experimental/routes_refactor directory. I was thinking once this feature is the default, we could get rid of this directory altogether. Let me know if you prefer we go a different route.

**Review by shashankram — APPROVED** [review]

File: experimental/routes_refactor/demo/scripts/port-forward-bookbuyer-envoy.sh (hunk 2 above)
Comment (snehachhabria) [reply experimental/routes_refactor/demo/scripts/port-forward-bookbuyer-envoy.sh:1]: I didn't notice the version, I think its okay to keep this directory for now.

File: pkg/envoy/rds/new_response.go (hunk 3 above)
Comment (michelleN) [reply pkg/envoy/rds/new_response.go]: updated. thank you

File: pkg/envoy/rds/new_response.go (hunk 4 above)
Comment (michelleN) [reply pkg/envoy/rds/new_response.go:28]: will add one

File: pkg/trafficpolicy/trafficpolicy.go (hunk 5 above)
Comment (michelleN) [reply pkg/trafficpolicy/trafficpolicy.go:37]: will add one now. thanks

**Review by snehachhabria — APPROVED** [review]

**Pull request merged**